BITSWAP.md

```diff
@@ -79,122 +75,126 @@ message Message {
+prefixes and block data. The cid prefixes are used to ensure the correct
+codecs and hash functions are used to handle the block on the receiving
+end.
 
-The network is the abstraction representing all Bitswap peers that are connected to us by one or more hops. Bitswap messages flow in and out of the network. This is where a game-theoretical analysis of Bitswap becomes relevant – in an arbitrary network we must assume that all of our peers are rational and self-interested, and we act accordingly. Work along these lines can be found in the [research-bitswap repository](https://github.com/ipfs/research-bitswap), with a preliminary game-theoretical analysis currently in-progress [here](https://github.com/ipfs/research-bitswap/blob/docs/strategy_analysis/analysis/prelim_strategy_analysis.pdf).
+It is otherwise identical to 1.0.0
 
-# Implementation Details
+### Wire Format
 
-## Coalescing Messages
-
-When a message queue that already contains a Bitswap message receives another, the new message should be coalesced with the original to reduce the overhead of sending separate packets.
-
-## Bitswap Sessions
-
-Bitswap sessions are an attempt to optimize the block requests sent to other Bitswap clients. When requesting a graph of blocks from the network, we send a wantlist update containing the graph’s root block to all of our peers. Then, for each peer who sends the root block back, we add that peer to the graph’s *active set*. We then send all requests for other nodes in the graph only to the peers in the active set. The idea is that peers who have the root node of a graph are likely to have its children as well, while those who do not have the root are unlikely to have its children.
+```
+message Message {
 
-**TODO**: Everything below must either be updated/integrated above, or removed
+  message Wantlist {
+    enum WantType {
+      Block = 0;
+      Have = 1;
+    }
 
-Also, make sure to read - <https://github.com/ipfs/go-ipfs/tree/master/exchange/bitswap#go-ipfs-implementation>
+    message Entry {
+			bytes block = 1; // CID of the block
+			int32 priority = 2;	// the priority (normalized). default to 1
+			bool cancel = 3; // whether this revokes an entry
+		}
 
-Implementation suggestions:
+    repeated Entry entries = 1;	// a list of wantlist entries
+    bool full = 2; // whether this is the full wantlist. default to false
+  }
 
-  - maintain a peer set of “live partners”
-  - protocol listener accept streams for partners to receive messages
-  - protocol sender opens streams to partners to send messages
-  - separate out a decision engine that selects which blocks to send to which partners, and at what time. (this is a bit tricky, but it’s super easy to make as a whole if the engine is separated out)
+  message Block {
+    bytes prefix = 1; // CID prefix (all of the CID components except for the digest of the multihash)
+    bytes data = 2;
+  }
 
-Sender:
+  Wantlist wantlist = 1;
+  repeated Block payload = 3; 
```

Comment:
While it appears to be supported in go-bitswap I removed `repeated bytes blocks = 2;` from versions of the protocol >1.0.0 as that field is deprecated and superseded by another field (i.e. `payload`). Given that we use libp2p and multistream to negotiate protocol versions there seems to be no reason why a `/ipfs/bitswap/1.1.0` node should ever receive a message with the `blocks` field since a node needing the `blocks` field should have negotiated for `/ipfs/bitswap/1.0.0`.

However, given that this behavior is likely implemented across the various implementations today I'm happy to put it back in the spec if people think it should be there.